### PR TITLE
Add svm bid data extraction from transaction

### PR DESCRIPTION
**Pull request by danimhr**
This PR aims to extract bid data from svm transaction

### CODE REVIEW

File: auction-server/src/auction.rs [hunk 1]
```diff
@@ -895,15 +908,48 @@ pub fn verify_submit_bid_instruction_svm(
...
+        _ => Err(RestError::BadParameters(
             "Bid has to include exactly one submit_bid instruction to Express Relay program"
                 .to_string(),
-        ));
+        )),
     }
-    Ok(())
+}
+
+fn find_account_position_svm(idl: Idl, instruction: &str, account: &str) -> Option<usize> {
```
Comment (m30m) [inline auction-server/src/auction.rs:923]: I'd rather find the index in build.rs and pass it here as a constant instead of finding it here again

File: auction-server/src/auction.rs [hunk 2]
```diff
@@ -895,15 +908,48 @@ pub fn verify_submit_bid_instruction_svm(
...
+    let discriminator = express_relay_svm::instruction::SubmitBid::discriminator();
+    let submit_bid_data = express_relay_svm::SubmitBidArgs::try_from_slice(
+        &instruction.data.as_slice()[discriminator.len()..],
+    )
+    .map_err(|e| RestError::BadParameters(format!("Invalid submit_bid instruction data: {}", e)))?;
+    let permission_account_position =
+        find_account_position_svm(idl, SUBMIT_BID_INSTRUCTION_SVM, PERMISSION_ACCOUNT_SVM).ok_or(
+            RestError::BadParameters("Invalid submit_bid instruction accounts".to_string()),
+        )?;
+    let account_position = instruction.accounts[permission_account_position];
```
Comment (m30m) [inline auction-server/src/auction.rs:948]: this can lead to out of bound panics

File: auction-server/src/auction.rs [hunk 3]
```diff
@@ -895,15 +908,48 @@ pub fn verify_submit_bid_instruction_svm(
...
+    Ok((
+        submit_bid_data.bid_amount,
+        accounts[account_position as usize],
```
Comment (m30m) [inline auction-server/src/auction.rs:951]: not sure, but this may also lead to out of bound panics

**Review by m30m — APPROVED** [review]

File: auction-server/src/auction.rs [hunk 4]
```diff
@@ -895,15 +908,48 @@ pub fn verify_submit_bid_instruction_svm(
...
+    idl: Idl,
+    accounts: &[Pubkey],
+    instruction: CompiledInstruction,
+) -> Result<(u64, Pubkey), RestError> {
+    let discriminator = express_relay_svm::instruction::SubmitBid::discriminator();
+    let submit_bid_data = express_relay_svm::SubmitBidArgs::try_from_slice(
+        &instruction.data.as_slice()[discriminator.len()..],
+    )
+    .map_err(|e| RestError::BadParameters(format!("Invalid submit_bid instruction data: {}", e)))?;
+    let permission_account_position =
```
Comment (anihamde) [inline auction-server/src/auction.rs:944]: i would prefer you also get the router account and index by that. generally, permission keys _should_ be unique across different router opportunities, but the way the program is structured, permissioning is checked for a tuple `(permission, router)`. it would be more accurate to also extract the router account as part of the bid data

File: auction-server/src/auction.rs [hunk 5]
```diff
@@ -895,15 +908,48 @@ pub fn verify_submit_bid_instruction_svm(
...
+}
+
+const SUBMIT_BID_INSTRUCTION_SVM: &str = "submit_bid";
+const PERMISSION_ACCOUNT_SVM: &str = "permission";
+
+fn extract_bid_data_svm(
+    idl: Idl,
+    accounts: &[Pubkey],
+    instruction: CompiledInstruction,
+) -> Result<(u64, Pubkey), RestError> {
```
Comment (anihamde) [inline auction-server/src/auction.rs:938]: i think you also should extract the bid `deadline`? this could be relevant for deciding whether to include this bid in the auction or not (if deadline has already passed)

File: auction-server/src/auction.rs [hunk 6]
```diff
@@ -895,15 +908,48 @@ pub fn verify_submit_bid_instruction_svm(
...
+    let discriminator = express_relay_svm::instruction::SubmitBid::discriminator();
+    let submit_bid_data = express_relay_svm::SubmitBidArgs::try_from_slice(
+        &instruction.data.as_slice()[discriminator.len()..],
+    )
+    .map_err(|e| RestError::BadParameters(format!("Invalid submit_bid instruction data: {}", e)))?;
+    let permission_account_position =
+        find_account_position_svm(idl, SUBMIT_BID_INSTRUCTION_SVM, PERMISSION_ACCOUNT_SVM).ok_or(
```
Comment (anihamde) [inline auction-server/src/auction.rs:945]: i think you can probably run this method once upon startup and not have to run it each time?

File: auction-server/src/auction.rs [hunk 7]
```diff
@@ -895,15 +908,48 @@ pub fn verify_submit_bid_instruction_svm(
...
+        IdlInstructionAccountItem::Single(a) => a.name == account,
+        IdlInstructionAccountItem::Composite(a) => a.name == account,
+    })
+}
+
+const SUBMIT_BID_INSTRUCTION_SVM: &str = "submit_bid";
+const PERMISSION_ACCOUNT_SVM: &str = "permission";
+
+fn extract_bid_data_svm(
+    idl: Idl,
```
Comment (anihamde) [inline auction-server/src/auction.rs:935]: if you run the account index method once upon startup, you don't need to pass the IDL here

File: auction-server/src/auction.rs (hunk 5 above)
Comment (danimhr) [reply auction-server/src/auction.rs:923]: We will run a simulation after the extraction. If the deadline is invalid, then the simulation will be failed.

File: auction-server/src/auction.rs (hunk 4 above)
Comment (danimhr) [reply auction-server/src/auction.rs]: Is there any chance that the permission be same for two different opportunity (auction)?

File: auction-server/src/server.rs [hunk 8]
```diff
@@ -313,6 +313,9 @@ pub async fn start_server(run_options: RunOptions) -> anyhow::Result<()> {
         access_tokens: RwLock::new(access_tokens),
         metrics_recorder: setup_metrics_recorder()?,
         express_relay_idl: load_express_relay_idl()?,
```
Comment (m30m) [inline auction-server/src/server.rs]: I don't think we need this anymore

File: auction-server/src/auction.rs [hunk 9]
```diff
@@ -895,15 +904,53 @@ pub fn verify_submit_bid_instruction_svm(
...
+    accounts: &[Pubkey],
+    instruction: CompiledInstruction,
+) -> Result<(u64, Pubkey), RestError> {
+    let discriminator = express_relay_svm::instruction::SubmitBid::discriminator();
+    let submit_bid_data = express_relay_svm::SubmitBidArgs::try_from_slice(
+        &instruction.data.as_slice()[discriminator.len()..],
+    )
+    .map_err(|e| RestError::BadParameters(format!("Invalid submit_bid instruction data: {}", e)))?;
+
+    if instruction.accounts.len() <= permission_account_position {
```
Comment (m30m) [inline auction-server/src/auction.rs]: I think it's cleaner to use `instruction.accounts.get` instead of manual out of bound check

File: auction-server/src/auction.rs [hunk 10]
```diff
@@ -895,15 +904,53 @@ pub fn verify_submit_bid_instruction_svm(
...
+            permission_account_position,
+        );
+        return Err(RestError::BadParameters(
+            "Permission account not found in submit_bid instruction".to_string(),
         ));
     }
-    Ok(())
+
+    let account_position = instruction.accounts[permission_account_position] as usize;
+    if account_position >= accounts.len() {
```
Comment (m30m) [inline auction-server/src/auction.rs:938]: use `accounts.get` instead of manual check

File: auction-server/src/auction.rs [hunk 11]
```diff
@@ -895,15 +904,53 @@ pub fn verify_submit_bid_instruction_svm(
...
+            "Permission account not found in transaction accounts: {:?} - {}",
+            accounts,
+            account_position,
+        );
+        return Err(RestError::BadParameters(
+            "Permission account not found in transaction accounts".to_string(),
+        ));
+    }
+
+    Ok((submit_bid_data.bid_amount, accounts[account_position]))
```
Comment (m30m) [inline auction-server/src/auction.rs]: we should do bound check here

File: auction-server/build.rs [hunk 12]
```diff
@@ -1,13 +1,23 @@
...
+};
 
-fn main() {
+fn build_evm_contracts() {
     let contract_setup = r#"
         cd ../contracts/evm
         forge build --via-ir
     "#;
     println!("cargo:rerun-if-changed=../contracts/evm");
-    println!("cargo:rerun-if-changed=migrations");
```
Comment (m30m) [inline auction-server/build.rs:9]: why is this removed? I added this because of this:
https://docs.rs/sqlx/latest/sqlx/macro.migrate.html

File: auction-server/src/server.rs [hunk 13]
```diff
@@ -267,6 +268,15 @@ pub async fn start_server(run_options: RunOptions) -> anyhow::Result<()> {
     };
 
     let chains_svm = setup_chain_store_svm(config_map);
+    let express_relay_svm = ExpressRelaySvm {
+        idl:                         load_express_relay_idl()?,
```
Comment (m30m) [inline auction-server/src/server.rs:272]: we never use this

File: auction-server/src/server.rs [hunk 14]
```diff
@@ -267,6 +268,15 @@ pub async fn start_server(run_options: RunOptions) -> anyhow::Result<()> {
...
+        permission_account_position: env!("SUBMIT_BID_PERMISSION_ACCOUNT_POSITION")
```
Comment (m30m) [inline auction-server/src/server.rs:273]: I'd rather put this inside the function that is using it. Soon, we can make these constant and fully compile time:
https://github.com/rust-lang/rust/pull/124941

File: auction-server/src/auction.rs [hunk 15]
```diff
@@ -895,15 +906,73 @@ pub fn verify_submit_bid_instruction_svm(
...
+            instruction,
+            position,
+        );
+        return Err(RestError::BadParameters(
+            "Permission account not found in submit_bid instruction".to_string(),
         ));
     }
-    Ok(())
+
+    let account_position = instruction.accounts[position] as usize;
```
Comment (m30m) [inline auction-server/src/auction.rs:937]: I think using `into` is safer than `as`:
`let account_position: usize = instruction.accounts[position].into();`

File: auction-server/src/auction.rs [hunk 16]
```diff
@@ -895,15 +906,73 @@ pub fn verify_submit_bid_instruction_svm(
...
+        )),
+    }
+}
+
+fn extract_account_svm(
+    accounts: &[Pubkey],
+    instruction: CompiledInstruction,
+    position: usize,
+) -> Result<Pubkey, RestError> {
+    if instruction.accounts.len() <= position {
```
Comment (m30m) [inline auction-server/src/auction.rs:926]: use `accounts.get` instead of manual boundry check

File: auction-server/build.rs (hunk 12 above)
Comment (danimhr) [reply auction-server/build.rs:9]: It isn't removed. It's at the beginning of the `main` function

File: auction-server/src/server.rs (hunk 14 above)
Comment (danimhr) [reply auction-server/src/server.rs:273]: Why not here?
If the value doesn't exist, then the run will be failed.

File: auction-server/build.rs (hunk 12 above)
Comment (m30m) [reply auction-server/build.rs:9]: my bad

File: auction-server/src/server.rs (hunk 14 above)
Comment (m30m) [reply auction-server/src/server.rs:261]: fair enough, but hopefully we can get rid of this struct later on when we can make these const values
